File: frontend/src/lib/zod/enrollment-letter.ts
// src/lib/zod/enrollment-letter.ts
//
// Mirror của `EnrollmentLetterResponse` (Backend_FastAPI/app/schemas/
// enrollment_letter.py). Danh sách bản đã phát là dữ liệu của một VĂN BẢN
// CHÍNH THỨC, nên parse ở runtime thay vì chỉ khai interface: backend đổi/bỏ
// field thì TypeScript vẫn biên dịch sạch với interface cũ và UI lặng lẽ hiện
// `undefined` — đúng thứ frontend/CLAUDE.md cấm ("Require server to provide").

import { z } from "zod"

export const enrollmentLetterSummarySchema = z.object({
  id: z.number(),
  profile_id: z.number(),
  /** ISO date (YYYY-MM-DD) — cửa sổ làm thủ tục in trên giấy. */
  enrollment_start_date: z.string(),
  enrollment_end_date: z.string(),
  generated_by_id: z.number().nullable(),
  /** ISO datetime. */
  generated_at: z.string(),
  file_size: z.number(),
  /** Hết hạn lưu trữ; null = không đặt hạn. */
  expires_at: z.string().nullable(),
  /** null = BẢN HIỆN HÀNH; có giá trị = đã bị bản phát sau thay thế. */
  superseded_at: z.string().nullable(),
})

export const enrollmentLetterListSchema = z.array(enrollmentLetterSummarySchema)

export type EnrollmentLetterSummary = z.infer<
  typeof enrollmentLetterSummarySchema
>

File: Backend_FastAPI/app/schemas/enrollment_letter.py
# app/schemas/enrollment_letter.py
"""Pydantic schemas for the official "Giấy báo nhập học" issuance."""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class EnrollmentLetterIssueRequest(BaseModel):
    """Body for POST issue — the officer-entered enrollment window."""

    enrollment_start_date: date = Field(
        ..., description="Ngày bắt đầu lên trường làm thủ tục"
    )
    enrollment_end_date: date = Field(
        ..., description="Ngày kết thúc (phải >= ngày bắt đầu)"
    )

    @model_validator(mode="after")
    def _validate_range(self) -> "EnrollmentLetterIssueRequest":
        if self.enrollment_end_date < self.enrollment_start_date:
            raise ValueError("Ngày kết thúc phải lớn hơn hoặc bằng ngày bắt đầu.")
        return self


class EnrollmentLetterResponse(BaseModel):
    """Issued-letter metadata (for the list / re-download UI). No PII payload."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    profile_id: int
    enrollment_start_date: date
    enrollment_end_date: date
    generated_by_id: Optional[int]
    generated_at: datetime
    file_size: int
    expires_at: Optional[datetime]
    # NULL = bản hiện hành. FE dùng để phân biệt "bản đang có hiệu lực" với
    # "bản đã bị thay thế" — N bản đều có chữ ký nên không đánh dấu thì không
    # ai biết bản nào là bản đã trao cho thí sinh.
    superseded_at: Optional[datetime]
    # Còn tải lại được không. BE trả lời sẵn để FE KHÔNG phải tự so ngày:
    # so ngày ở client vừa trái Thin Client vừa sai khi đồng hồ máy officer
    # lệch, mà bản đã bị retention xoá file thì `expires_at` cũng không đủ để
    # suy ra (row vẫn còn hạn nhưng file đã biến mất).
    is_downloadable: bool
